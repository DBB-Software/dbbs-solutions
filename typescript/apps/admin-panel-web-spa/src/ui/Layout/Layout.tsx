import React, { FC, ReactNode, useState } from 'react'
import { AppBar, CssBaseline, Drawer, IconButton, Box, useMediaQuery, Typography, Divider } from '@dbbs/mui-components'
import { Link, useRouterState } from '@tanstack/react-router'
import { Menu } from 'lucide-react'
import { SIDEBAR_DATA_PAGES_LINKS, makeSxStyles } from '../../utils'
import theme from '../../theme'
import { SidebarLink } from '../SidebarLink'

export type LayoutProps = {
  children?: ReactNode
}

const makeStyles = () =>
  makeSxStyles({
    root: {
      display: 'flex'
    },
    header: {
      position: 'fixed',
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'flex-start',
      alignItems: 'center',
      backgroundColor: 'white',
      px: 2,
      height: 48,
      zIndex: theme.zIndex.drawer + 1
    },
    menuButton: { mr: 2 },
    sidebarRoot: {
      width: { md: 240 },
      flexShrink: { md: 0 }
    },
    sidebarDrawer: {
      display: { xs: 'none', md: 'block' },
      '& .MuiDrawer-paper': {
        minWidth: 240
      }
    },
    sidebarMobileDrawer: {
      display: { xs: 'block', md: 'none' },
      '& .MuiDrawer-paper': { minWidth: 240, pt: 6 }
    },
    contentWrapper: {
      flexGrow: 1,
      p: 3,
      mt: {
        md: 0,
        xs: 6
      },
      width: 'auto'
    },
    menuContentWrapper: {
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'space-between',
      flexGrow: 1
    },
    navigationLink: {
      textDecoration: 'none',
      fontWeight: 'bold',
      color: theme.palette.primary
    },
    contentRoot: {
      flexGrow: 1,
      p: 2,
      display: 'flex',
      flexDirection: 'column',
      gap: 2
    },
    logo: {
      textDecoration: 'none'
    }
  })

export const Layout: FC<LayoutProps> = ({ children }) => {
  const styles = makeStyles()
  const isMobile = useMediaQuery(theme.breakpoints.down('md'))

  const [mobileOpen, setMobileOpen] = useState(false)

  const router = useRouterState()

  const handleDrawerToggle = () => {
    setMobileOpen(!mobileOpen)
  }

  const drawerContent = (
    <Box sx={styles.contentRoot}>
      <Link to="/" style={{ textDecoration: 'none', color: 'inherit' }}>
        <Typography typography="h3" alignContent="center">
          LOGO
        </Typography>
      </Link>
      <Divider />

      <Box sx={styles.menuContentWrapper}>
        <Box>
          {SIDEBAR_DATA_PAGES_LINKS.map((linkItem) => (
            <SidebarLink {...linkItem} isActive={router.location.pathname === linkItem.link} key={linkItem.key} />
          ))}
        </Box>
      </Box>
    </Box>
  )

  return (
    <Box sx={styles.root}>
      <CssBaseline />
      {isMobile && (
        <AppBar sx={styles.header}>
          <IconButton edge="start" onClick={handleDrawerToggle} sx={styles.menuButton}>
            <Menu />
          </IconButton>
        </AppBar>
      )}

      <Box sx={styles.sidebarRoot}>
        <Drawer
          variant={isMobile ? 'temporary' : 'permanent'}
          open={!isMobile || mobileOpen}
          onClose={handleDrawerToggle}
          ModalProps={{ keepMounted: true }}
          sx={styles.sidebarMobileDrawer}
        >
          {drawerContent}
        </Drawer>

        <Drawer variant="permanent" sx={styles.sidebarDrawer} open>
          {drawerContent}
        </Drawer>
      </Box>

      <Box component="main" sx={styles.contentWrapper}>
        {children}
      </Box>
    </Box>
  )
}
