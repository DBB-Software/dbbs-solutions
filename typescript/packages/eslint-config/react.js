const tsConfigs = ['tsconfig.json', 'tsconfig.spec.json', 'tsconfig.app.json', 'tsconfig.lib.json']

module.exports = {
  env: {
    browser: true,
    node: true,
    jest: true,
    'cypress/globals': true
  },
  extends: [
    'airbnb-base',
    'airbnb-typescript',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:react-hooks/recommended',
    'plugin:react/recommended',
    'eslint:recommended',
    'prettier',
    'turbo'
  ],
  parserOptions: {
    project: tsConfigs
  },
  plugins: ['prettier', '@typescript-eslint', 'jest', 'cypress'],
  rules: {
    'turbo/no-undeclared-env-vars': 'off',
    'no-unused-vars': 'off',
    'react/react-in-jsx-scope': 'off',
    'no-console': 'off',
    semi: ['error', 'never'],
    'import/prefer-default-export': 'off',
    'import/no-extraneous-dependencies': 'off',
    '@typescript-eslint/no-unused-vars': 'error',
    '@typescript-eslint/ban-types': 'warn',
    '@typescript-eslint/ban-ts-comment': 'warn',
    '@typescript-eslint/no-explicit-any': 'warn',
    'react/function-component-definition': [
      'warn',
      {
        namedComponents: ['function-declaration', 'function-expression', 'arrow-function'],
        unnamedComponents: ['function-expression', 'arrow-function']
      }
    ],
    'react/jsx-props-no-spreading': 'off',
    'react-hooks/rules-of-hooks': 'error', // Checks rules of Hooks
    'react-hooks/exhaustive-deps': 'warn', // Checks effect dependencies
    'prefer-const': 'error',
    'no-import-assign': 'error',
    'no-self-assign': 'error',
    'jsx-quotes': ['error', 'prefer-double'],
    'no-restricted-exports': 'off',
    'import/extensions': 'off',
    'array-callback-return': 'off',
    'class-methods-use-this': 'off',
    'consistent-return': 'off',
    'react/display-name': 'off',
    quotes: ['error', 'single'],
    'prettier/prettier': [
      'error',
      {
        endOfLine: 'auto'
      }
    ],
    'react/prop-types': 'off'
  },
  settings: {
    react: {
      version: 'detect'
    }
  },
  ignorePatterns: ['node_modules/', 'dist/', 'lib/', 'coverage/', 'build/']
}