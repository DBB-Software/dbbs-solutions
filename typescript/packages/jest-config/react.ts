import type { Config } from 'jest'
import { baseConfig } from './base.js'

export const reactConfig: Config = {
  ...baseConfig,
  displayName: 'dbbs-web-spa',
  testEnvironment: 'jsdom',
  transformIgnorePatterns: ['node_modules/'],
  setupFilesAfterEnv: ['./src/testUtils/setupTests.ts']
}
