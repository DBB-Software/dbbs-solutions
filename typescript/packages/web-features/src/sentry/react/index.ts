export * from './sentry'

export { ErrorBoundary } from './ErrorBoundary'
