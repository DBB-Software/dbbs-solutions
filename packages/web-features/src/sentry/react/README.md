<p align="center">
  <a href="https://sentry.io/?utm_source=github&utm_medium=logo" target="_blank">
    <img src="https://sentry-brand.storage.googleapis.com/sentry-wordmark-dark-280x84.png" alt="Sentry" width="280" height="84">
  </a>
</p>

# DBB wrapper for Sentry SDK for ReactJS

## Links

- [Official SDK Docs](https://docs.sentry.io/platforms/javascript/guides/react/)
- [TypeDoc](http://getsentry.github.io/sentry-javascript/)

## General

This package is a wrapper around `@sentry/react`, with added functionality related to React.

To use this SDK, call `Sentry.init(options)` before you mount your React component.

Remember to add the environment variable to your .env file.

```env
 WEB_APP_SENTRY_DSN=
```

```javascript
import React from 'react';
import { createRoot } from 'react-dom/client';
import { init } from '@dbbs/web-features/src/sentry/react'

init()

// ...

const container = document.getElementById(“app”);
const root = createRoot(container);
root.render(<App />);

// also works with hydrateRoot
// const domNode = document.getElementById('root');
// const root = hydrateRoot(domNode, reactNode);
// root.render(<App />);
```

### ErrorBoundary

`@dbbs/web-features/src/sentry/react` exports an ErrorBoundary component that will automatically send Javascript errors from inside a
component tree to Sentry, and set a fallback UI.

> app.js

```javascript
import React from 'react';
import { ErrorBoundary } from '@dbbs/web-features/src/sentry/react'

function FallbackComponent() {
  return <div>An error has occured</div>;
}

class App extends React.Component {
  render() {
    return (
      <ErrorBoundary fallback={FallbackComponent} showDialog>
        <OtherComponents />
      </ErrorBoundary>
    );
  }
}

export default App;
```
