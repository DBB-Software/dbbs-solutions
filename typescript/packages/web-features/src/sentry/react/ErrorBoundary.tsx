import { FC } from 'react'
import * as Sentry from '@sentry/react'

/**
 * ErrorBoundary component that wraps Sentry's ErrorBoundary with the provided props.
 *
 * @param {Sentry.ErrorBoundaryProps} props - Props passed to Sentry's ErrorBoundary component.
 * @returns {JSX.Element} A Sentry ErrorBoundary component with the provided props.
 */
export const ErrorBoundary: FC<Sentry.ErrorBoundaryProps> = (props) => <Sentry.ErrorBoundary {...props} />
