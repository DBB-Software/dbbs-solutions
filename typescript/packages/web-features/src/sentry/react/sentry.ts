import * as Sentry from '@sentry/react'

/**
 * Initializes Sentry with the provided options or environment variables.
 * The DSN is retrieved from the environment variable WEB_APP_SENTRY_DSN.
 *
 * @param {Parameters<typeof Sentry.init>[0]} [options={}] - Configuration options for Sentry initialization.
 */
export const init = (options: Parameters<typeof Sentry.init>[0] = {}) => {
  Sentry.init({
    dsn: process?.env?.WEB_APP_SENTRY_DSN,
    ...options
  })
}
/**
 * Exporting various Sentry methods for capturing and setting different types of data.
 */
export {
  setExtra,
  setTag,
  setUser,
  addBreadcrumb,
  captureEvent,
  captureMessage,
  captureSession,
  captureException
} from '@sentry/react'
