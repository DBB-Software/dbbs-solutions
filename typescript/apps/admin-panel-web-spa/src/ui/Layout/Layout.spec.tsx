import { render } from '@testing-library/react'
import React from 'react'
import { Layout } from './Layout'

jest.mock('@tanstack/react-router', () => ({
  ...jest.requireActual('@tanstack/react-router'),
  useRouterState: jest.fn(() => ({
    location: { pathname: '/' }
  })),
  Link: jest.fn()
}))

jest.mock('@dbbs/mui-components', () => ({
  ...jest.requireActual('@dbbs/mui-components'),
  useMediaQuery: jest.fn(() => true)
}))

describe('<Layout />', () => {
  it('renders correctly', async () => {
    const { asFragment } = render(<Layout />)

    expect(asFragment()).toMatchSnapshot()
  })
})
