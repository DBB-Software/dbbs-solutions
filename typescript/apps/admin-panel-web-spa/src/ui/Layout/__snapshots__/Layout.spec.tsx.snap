// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-k008qs"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-vongwo-MuiPaper-root-MuiAppBar-root"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeStart MuiIconButton-sizeMedium css-7ubvsf-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          class="lucide lucide-menu"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="4"
            x2="20"
            y1="12"
            y2="12"
          />
          <line
            x1="4"
            x2="20"
            y1="6"
            y2="6"
          />
          <line
            x1="4"
            x2="20"
            y1="18"
            y2="18"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </header>
    <div
      class="MuiBox-root css-1at9qkq"
    >
      <div
        class="MuiDrawer-root MuiDrawer-docked css-4u0zen-MuiDrawer-docked"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft css-12i7wg6-MuiPaper-root-MuiDrawer-paper"
        >
          <div
            class="MuiBox-root css-5tys7z"
          >
            <hr
              class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
            />
            <div
              class="MuiBox-root css-a74d04"
            >
              <div
                class="MuiBox-root css-0"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <main
      class="MuiBox-root css-12rwh79"
    />
  </div>
</DocumentFragment>
`;
